server/path_watcher.go

```diff
@@ -0,0 +1,309 @@
+// Copyright (C) 2019 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cloudwan/gohan/extension"
+	"github.com/cloudwan/gohan/extension/goext"
+	l "github.com/cloudwan/gohan/log"
+	"github.com/cloudwan/gohan/metrics"
+	gohan_sync "github.com/cloudwan/gohan/sync"
+	"github.com/cloudwan/gohan/util"
+	"github.com/pkg/errors"
+)
+
+type PathWatcher struct {
+	sync                      gohan_sync.Sync
+	priority                  int
+	path                      string
+	escapedPath               string
+	extensions                map[string]extension.Environment
+	previousProcessedRevision int64
+}
+
+var (
+	errInconsistentCluster = errors.New("inconsistent cluster state detected")
+	replacer               = strings.NewReplacer(".", "_", "/", "_")
+)
+
+func NewPathWatcher(sync gohan_sync.Sync, extensions map[string]extension.Environment, path string, priority int) *PathWatcher {
+	return &PathWatcher{
+		sync:        sync,
+		extensions:  extensions,
+		priority:    priority,
+		path:        path,
+		escapedPath: replacer.Replace(path),
+	}
+}
+
+func (watcher PathWatcher) String() string {
+	sb := strings.Builder{}
+	sb.WriteString("(PathWatcher ")
+	sb.WriteString(watcher.path)
+	sb.WriteString(")")
+	return sb.String()
+}
+
+func (watcher *PathWatcher) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	select {
+	case <-time.After(time.Duration(watcher.priority*masterTTL) * time.Second):
+	case <-ctx.Done():
+		return
+	}
+
+	for {
+		err := watcher.run(ctx)
+
+		switch err {
+		case errLockFailed:
+			watcher.updateCounter(1, "lock.failed")
+			log.Debug("%s failed to acquire lock, retrying...", watcher)
```

Comment:
it's because `log` refers to a package-scope [variable](https://github.com/cloudwan/gohan/blob/CW-1065-fix-sync-watcher-races/server/log.go#L22), not a package name

Reply:
Found it. Package `server` defines `var log = l.NewLogger()` and `path_watcher.go` happens to belong to that package. It seems we have 2 methods of logging in this file...

Reply:
Scratch all that, `l.Panic(log)` is actually panic-recovery, name not misleading at all.

Reply:
I had to lookup it too during first glance. Can you change name to `LogPanic(log)`? Also 'l' is declared too far from scope where its used.

Reply:
It's kind of funny that `Panic(log)` does something entirely different than `log.Panic()` ;)

Reply:
we can do better: the [util.go](https://github.com/cloudwan/gohan/blob/master/log/util.go) had two functions:

1. Panic() which is incorrectly named and is called only here, and
2. FatalPanic() which is called just once (in cron.go), but it's call both redundantly & incorrectly (it's actually a dead code)

let me sort it out for good.